Pairwise Intersections

ayush1999 opened the pull request:
#12949 

Added a "union" argument. 

- **When union is False**, the function works just like before.

- **When union is True** , the function returns all points of intersections.

ayush1999 committed:
Pairwise Intersections

asmeurer commented:
Can you indicate what this does more clearly? A doctest would be best. Also this needs real tests. 

From what I can tell, `intersection` already returns all the points of intersection, so what is the difference? Also, to me, `union` is the opposite of `intersection`, so it's a very confusing name.

ayush1999 commented:
@asmeurer . This was raised in the issue #12949 

I have added a new argument called union.
The conditions are:

- When union is False, the functions works as it used to, and **returns the points through which all the entities pass.**

-When union is True, **the function returns all points of intersection ( these points may not necessarily lie on all the entities ) This means that all points where pairwise intersection of entities takes place is also returned.**

This makes the intersection function more universal.

ayush1999 commented:
@asmeurer Since the function now returns pairwise intersections , I though that **union** could be a suitable name for the argument.

asmeurer commented:
I see. `pairwise=True` would be a better keyword name.

ayush1999 commented:
@asmeurer Should I change the name?

asmeurer commented:
I think it should be. Maybe @smichr also has some thoughts.

smichr commented:
You might do a text search to see if pairwise or union are already used as keywords. `pairwise` seems ok, too.

ayush1999 committed:
Changed argument name to: pairwise

ayush1999 committed:
Error fixed.

ayush1999 commented:
@asmeurer @smichr . I have changed the argument name to **pairwise** , and also fixed the errors.

smichr commented on sympy/geometry/util.py:
```diff
@@ -684,23 +681,44 @@ def intersection(*entities):
-        return []
-
-    # entities may be an immutable tuple
+    pairwise = kwargs.pop('union', False)
```
pop 'pairwise'

smichr commented on sympy/geometry/util.py:
```diff
@@ -684,23 +681,44 @@ def intersection(*entities):
+
+    if pairwise == False:
+
+        if len(entities) <= 1:
```
put this outside of `if not pairwise:`

smichr commented on sympy/geometry/util.py:
```diff
@@ -684,23 +681,44 @@ def intersection(*entities):
+        if len(entities) <= 1:
+            return []
+
+        # entities may be an immutable tuple
```
put this outside the if, too

smichr commented on sympy/geometry/util.py:
```diff
@@ -684,23 +681,44 @@ def intersection(*entities):
+        return res
+
+    ans = []
     entities = list(entities)
```
eliminate this since it will be handled about

smichr commented on sympy/geometry/util.py:
```diff
@@ -684,23 +681,44 @@ def intersection(*entities):
-    return res
+    for j in range(0, len(entities)):
+        for k in range(j + 1, len(entities)):
+            if len(intersection(entities[j], entities[k])) is not 0:
```
don't calculate intersection twice: capture then check

```
for j in range(len(entities)):
    for k in range(j + 1, len(entities)):
            ans.extend(intersection(entities[j], entities[k]))

return list(ordered(uniq(ans)))
```
```

smichr replied:
Maybe we should just have a new function, too, and make the keyword on intersection private (i.e. `_pairwise`) to discourage its use: 
```
def pairwise_intersections(*objs):
    """Return all unique pairwise intersections that occur between the given objects.
    return intersection(*objs, pairwise=True)
```

And here is a test:

```
c, l, s, t = Circle((0, 0), 1), Line((0, 0), slope=oo), Segment((0, 0), (0, 1)), Triangle((0, 1), (2, 0), (-2, 0))
assert pairwise_intersections(c, l, s, t) == [
    Point2D(-1, 0), Point2D(-4/5, 3/5), Point2D(0, 1), Point2D(4/5, 3/5), Point2D(1, 0), s]
```

ayush1999 replied:
Thank You @smichr . I'm working on it.

_Sent from my LENOVO Lenovo P1a42 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

ayush1999 committed:
Changes - #1

ayush1999 committed:
Changes- #2

ayush1999 commented:
@asmeurer @smichr I have made most of the required changes. Please review.

smichr commented on sympy/geometry/util.py:
```diff
@@ -684,23 +681,38 @@ def intersection(*entities):
-    # entities may be an immutable tuple
+    if not pairwise:
+
+        # entities may be an immutable tuple
```
This code converting tuples to Points needs to be moved out of the if-block. 

You can verify by using the test I gave before but put `(0,0)` as the first point -- no error should raise.

smichr committed:
factor out tuple->Point code; order results

smichr committed:
Update util.py

smichr commented:
I made some changes, but can't add the tests from here.

ayush1999 commented:
Also, @smichr , should the docstrings be added?

_Sent from my LENOVO Lenovo P1a42 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

ayush1999 committed:
Errors fixed : Trailing whitespace

ayush1999 committed:
Changes #3

ayush1999 committed:
Changes #4

smichr commented on sympy/geometry/util.py:
```diff
@@ -623,7 +623,6 @@ def idiff(eq, y, x, n=1):
         derivs[dydx] = yp
         eq = dydx - yp
         dydx = dydx.diff(x)
-  
```
There should be a double line space between functions at the main body level.

smichr commented on sympy/geometry/util.py:
```diff
@@ -623,7 +623,6 @@ def idiff(eq, y, x, n=1):
         dydx = dydx.diff(x)
-  
 def intersection(*entities, **kwargs):
     """The intersection of a collection of GeometryEntity instances.
```
The docstring should be ammended to reflect the new functionality.

ayush1999 replied:
@smichr I'll do that. Anything else to be done?

_Sent from my LENOVO Lenovo P1a42 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

ayush1999 committed:
Changes - #5

The following changes were made:
1. Docstrings were appended to include the new additions.
2. A few previous errors were fixed.
3. Newlines were added wherever necessary.

ayush1999 committed:
Errors fixed.

Indentation error occured by mistake

ayush1999 committed:
Newlines added

ayush1999 commented:
@asmeurer @smichr **All the necessary changes have been made**. These include:

- Adding the new argument and implementing it.

- Appending the docstrings to reflect this new functionality.

- Showing the working with the help of an example in the docstrings.

- Testing to make sure that this is giving correct results.

Please tell me if any extra thing needs to be done.
Also, Kindly check the changes and approve so that this PR can be merged.

smichr commented on sympy/geometry/util.py:
```diff
@@ -657,14 +654,21 @@ def intersection(*entities):
+
+    Case 2: When the keyword argument 'pairwise' is True:
+    In this case, the functions returns a list of all points
+    where intersections take place.(Please Note that these points
```
...where intersections take place.(Please Note that these points
may not necessarily lie on all the entities)

`-->`

...where intersections take place.  A given intersection may not necessarily
lie on all the entities.

You should probably avoid use of the word "point", too, since the intersection of two segments may be a segment.

smichr commented on sympy/geometry/util.py:
```diff
@@ -657,14 +654,21 @@ def intersection(*entities):
 
     Examples
     ========
-
```
unless something has changed, a blank line is required after Examples.

smichr commented on sympy/geometry/util.py:
```diff
@@ -657,14 +654,21 @@ def intersection(*entities):
     ========
-
     >>> from sympy.geometry import Point, Line, Circle, intersection
     >>> p1, p2, p3 = Point(0, 0), Point(1, 1), Point(-1, 5)
```
Let's make this example really simple to visualize and use the same example for both cases:

```
>>> from sympy import Cirlc, Line, Point, oo
>>> origin = Point(0, 0)
>>> c = Circle(origin, radius=1).translate(y=1)
>>> xaxis = Line(origin, slope=0)
>>> yaxis = Line(origin, slope=oo)
>>> intersection(c, xaxis, yaxis)
Point2D(0, 0)
>>> intersection(c, xaxis, yaxis, pairwise=True)
[Point2D(0, 0), Point2D(0, 2)]
```

smichr commented:
> any extra thing needs to be done.

I left a few comments. Also, a test in (probably) tests/test_util.py should be added. I would suggest adding one that involves segments, like

```
assert intersections(
    Segment((0, 0), (2, 0)), 
    Segment((-1, 0), (1, 0)), 
    Line((0,0), slope=oo), pairwise=True) == [
    Point(0, 0), Segment((0, 0), (1, 0))]
```

ayush1999 commented:
@smichr Shouldn't the test in tests/test_util.py be added after the PR is merged?

ayush1999 committed:
Changes #8

--> newlines added.
--> example changed

ayush1999 committed:
Errors fixed.

ayush1999 committed:
Error fixed

smichr committed:
simplify intersection docstring

Also show that an intersection can be more than a Point(s). Error handling is left to Point
to raise; NotImplementedError is not raised by Point.__new__.

ayush1999 commented:
@smichr @asmeurer I have made the required changes.

_Sent from my Lenovo P1a42 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

ayush1999 commented:
@smichr @asmeurer I think this can be merged now.

smichr commented:
See the comment above where I mention "test in (probably) tests/test_util.py." A test in the test suite needs to be added. The ones I mention there should be added; perhaps the one in the docstring that tests Segment output should be added, too.

ayush1999 commented:
@smichr I am having a little trouble with the function **ordered** that has been used in the last line.
Can you tell me where has this function been defined?

ayush1999 commented:
@smichr I have opened a new PR #13012 
**Three** test cases were added. In each, a small change has been done to show that the tests pass in all the possible cases ( as you had mentioned in the above comment )

ayush1999 committed:
Tests added to test_util.py

ayush1999 commented:
@smichr   **Three** test cases were added. In each, a small change has been done to show that the tests pass in all the possible cases ( as you had mentioned in the above comment )

ayush1999 committed:
Error fice. ( whitespace )

smichr committed:
Update test_util.py

smichr commented:
+1 when tests pass

smichr committed:
Update test_util.py
